blockchain/indexers/slpindex.go

```diff
@@ -0,0 +1,877 @@
+	// mint baton as burned.
+
+	// Check if tx is a valid SLP. SLP validity has two requirements:
+	//  (1) the slpMsg must be valid, and
+	//  (2) the input requirements must be satisfied.
+	isValid := false
+	outputAmt, _ := txSlpMsg.TotalSlpMsgOutputValue()
+	if txSlpMsg.TransactionType == "GENESIS" {
+		if txSlpMsg.TokenType == 0x41 &&
+			big.NewInt(1).Cmp(v1InputAmtSpent) < 1 {
+			isValid = true
+		} else if txSlpMsg.TokenType == 0x01 || txSlpMsg.TokenType == 0x81 {
+			isValid = true
+		}
+	} else if txSlpMsg.TransactionType == "SEND" &&
+		outputAmt.Cmp(v1InputAmtSpent) < 1 {
+		isValid = true
+	} else if txSlpMsg.TransactionType == "MINT" &&
+		v1MintBatonVout > 1 {
+		isValid = true
+	}
+
+	if isValid {
+		err := putTxIndexEntry(tx, txSlpMsg, tokenIDHash)
+		if err != nil {
+			panic(err.Error())
+		}
+	}
+	return isValid, burnedInputs
+}
+
+// DisconnectBlock is invoked by the index manager when a block has been
+// disconnected from the main chain.  This indexer removes the
+// hash-to-transaction mapping for every transaction in the block.
+//
+// This is part of the Indexer interface.
+func (idx *SlpIndex) DisconnectBlock(dbTx database.Tx, block *bchutil.Block, stxos []blockchain.SpentTxOut) error {
+
+	// Remove all of the transactions in the block from the index.
+	if err := dbRemoveSlpIndexEntries(dbTx, block); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GetSlpIndexEntry returns a serialized slp index entry for the provided transaction hash
+// from the slp index.  The slp index entry can in turn be used to quickly discover
+// additional slp information about the transaction. When there is no entry for the provided hash, nil
+// will be returned for the both the entry and the error, which would mean the transaction is invalid
+//
+// This function is safe for concurrent access.
+func (idx *SlpIndex) GetSlpIndexEntry(dbTx database.Tx, hash *chainhash.Hash) (*SlpIndexEntry, error) {
+	entry := idx.cache.Get(hash)
+	if entry != nil {
+		return entry, nil
+	}
+
+	entry, err := dbFetchSlpIndexEntry(dbTx, hash)
+	if err != nil {
+		return nil, err
+	}
+
+	idx.cache.AddTemp(hash, entry)
+	return entry, nil
+}
+
+// GetTokenMetadata ...
+func (idx *SlpIndex) GetTokenMetadata(dbTx database.Tx, tokenID uint32) (*TokenMetadata, error) {
+	serializedID := make([]byte, 4)
+	byteOrder.PutUint32(serializedID, tokenID)
+	return dbFetchTokenMetadataBySerializedID(dbTx, serializedID)
+}
+
+// AddMempoolTx adds a new SlpIndexEntry item to a temporary cache that holds
+// both mempool and recently queried db entries.
+//
+// TODO: How do we handle fetching of TokenMetadata for Genesis txns in the mempool?
+//
+func (idx *SlpIndex) AddMempoolTx(tx *bchutil.Tx) error {
+
+	slpMsg, err := v1parser.ParseSLP(tx.MsgTx().TxOut[0].PkScript)
+	if err != nil {
+		return err
+	}
+
+	if slpMsg.TokenType != 0x01 && slpMsg.TokenType != 0x41 && slpMsg.TokenType != 0x81 {
```

Comment:
Should we have an array or hash of accepted token types? Seems this logic is brittle and could easily be overlooked when adding a new type.

Reply:
I added a new method:

```
func isAcceptableTokenVersionType(versionType int) bool {
	if versionType == v1parser.TokenTypeFungible01 ||
		versionType == v1parser.TokenTypeNft1Child41 ||
		versionType == v1parser.TokenTypeNft1Group81 {
		return true
	}
	return false
}

```